// yakumo/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = exports.commands = exports.addHook = exports.hooks = exports.Project = exports.config = exports.requireSafe = exports.configRequire = exports.meta = exports.cwd = void 0;
const globby_1 = __importDefault(require("globby"));
const detect_indent_1 = __importDefault(require("detect-indent"));
const utils_1 = require("./utils");
const js_yaml_1 = require("js-yaml");
const fs_1 = require("fs");
const module_1 = require("module");
const cosmokit_1 = require("cosmokit");
__exportStar(require("./utils"), exports);
exports.cwd = process.cwd();
const content = (0, fs_1.readFileSync)(`${exports.cwd}/package.json`, 'utf8');
exports.meta = JSON.parse(content);
exports.configRequire = module_1.Module.createRequire(exports.cwd + '/package.json');
function requireSafe(id) {
    try {
        return (0, exports.configRequire)(id);
    }
    catch (e) {
        if (e.code !== 'MODULE_NOT_FOUND')
            throw e;
    }
}
exports.requireSafe = requireSafe;
function loadConfig() {
    let content;
    try {
        content = (0, fs_1.readFileSync)(`${exports.cwd}/yakumo.yml`, 'utf8');
    }
    catch { }
    return {
        alias: {},
        require: [],
        commands: {},
        ...content && (0, js_yaml_1.load)(content),
    };
}
exports.config = loadConfig();
class Project {
    constructor() {
        this.indent = (0, detect_indent_1.default)(content).indent;
        this.cwd = exports.cwd;
        this.config = exports.config;
        this.manager = utils_1.manager;
    }
    require(id) {
        return (0, exports.configRequire)(id);
    }
    async initialize(argv) {
        this.argv = argv;
        const folders = await (0, globby_1.default)(exports.meta.workspaces || [], {
            cwd: exports.cwd,
            onlyDirectories: true,
            expandDirectories: false,
        });
        folders.unshift('');
        this.workspaces = Object.fromEntries(folders.map((path) => {
            if (path)
                path = '/' + path;
            try {
                return [path, require(`${exports.cwd}${path}/package.json`)];
            }
            catch { }
        }).filter(Boolean));
        if (!argv._.length || argv.config.manual) {
            this.targets = { ...this.workspaces };
            return;
        }
        this.targets = (0, cosmokit_1.pick)(this.workspaces, argv._.flatMap((name) => {
            return this.locate(name);
        }));
    }
    locate(name) {
        if (exports.config.alias[name]) {
            return (0, cosmokit_1.makeArray)(exports.config.alias[name]).map((path) => {
                if (!this.workspaces[path]) {
                    throw new Error(`cannot find workspace ${path} resolved by ${name}`);
                }
                return path;
            });
        }
        const targets = Object.keys(this.workspaces).filter((folder) => {
            if (this.workspaces[folder].workspaces)
                return;
            const [last] = folder.split('/').reverse();
            return name === last;
        });
        if (!targets.length) {
            throw new Error(`cannot find workspace "${name}"`);
        }
        else if (targets.length > 1) {
            throw new Error(`ambiguous workspace "${name}": ${targets.join(', ')}`);
        }
        return targets;
    }
    async emit(name, ...args) {
        await Promise.all((exports.hooks[name] || []).map((callback) => callback.call(this, ...args)));
    }
    async save(path) {
        const content = JSON.stringify(this.workspaces[path], null, this.indent) + '\n';
        await fs_1.promises.writeFile(`${exports.cwd}${path}/package.json`, content);
    }
}
exports.Project = Project;
exports.hooks = {};
function addHook(name, callback) {
    (exports.hooks[name] || (exports.hooks[name] = [])).push(callback);
}
exports.addHook = addHook;
exports.commands = {};
function register(name, callback, options = {}) {
    exports.commands[name] = [callback, options];
}
exports.register = register;
