// yakumo/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.install = exports.manager = exports.spawnAsync = exports.exit = exports.confirm = void 0;
const ora_1 = __importDefault(require("ora"));
const prompts_1 = __importDefault(require("prompts"));
const which_pm_runs_1 = __importDefault(require("which-pm-runs"));
const execa_1 = __importDefault(require("execa"));
async function confirm(message) {
    const { value } = await (0, prompts_1.default)({
        name: 'value',
        type: 'confirm',
        message,
    });
    return value;
}
exports.confirm = confirm;
function exit(message) {
    const spinner = (0, ora_1.default)();
    spinner.info(message);
    return process.exit(0);
}
exports.exit = exit;
function spawnAsync(args, options = {}) {
    const child = (0, execa_1.default)(args[0], args.slice(1), options);
    return new Promise((resolve) => {
        var _a, _b;
        (_a = child.stderr) === null || _a === void 0 ? void 0 : _a.pipe(process.stderr);
        (_b = child.stdout) === null || _b === void 0 ? void 0 : _b.pipe(process.stdout);
        child.on('close', resolve);
    });
}
exports.spawnAsync = spawnAsync;
exports.manager = (0, which_pm_runs_1.default)();
async function install() {
    const agent = (exports.manager === null || exports.manager === void 0 ? void 0 : exports.manager.name) || 'npm';
    const code = await spawnAsync([agent, 'install']);
    if (code)
        process.exit(code);
}
exports.install = install;
